web/src/main/java/org/springframework/security/web/firewall/FirewalledResponse.java

```diff
@@ -34,10 +37,27 @@ public FirewalledResponse(HttpServletResponse response) {
 	public void sendRedirect(String location) throws IOException {
 		// TODO: implement pluggable validation, instead of simple blacklisting.
 		// SEC-1790. Prevent redirects containing CRLF
-		if (CR_OR_LF.matcher(location).find()) {
+		validateCRLF(location);
+		super.sendRedirect(location);
+	}
+
+	@Override
+	public void setHeader(String name, String value) {
+		validateCRLF(value);
+		super.setHeader(name, value);
+	}
+
+	@Override
+	public void addHeader(String name, String value) {
+		validateCRLF(value);
+		super.addHeader(name, value);
+	}
+
+	private void validateCRLF(String value) {
+		if (CR_OR_LF.matcher(value).find()) {
 			throw new IllegalArgumentException(
 					"Invalid characters (CR/LF) in redirect location");
```

Comment:
We need a different error message since these are not redirect locations. You could probably just change this to header since redirect location is a header.